# Data Preprocessing Pipeline
## Analisa Big Data - Rata-rata Pengeluaran Perkapita untuk Buah-buahan dan Sayur-sayuran

---

### 📊 **Dataset Overview**
Dataset ini berisi data rata-rata pengeluaran perkapita seminggu menurut kelompok buah-buahan dan sayur-sayuran per kabupaten/kota untuk tahun **2023** dan **2024**.

**Sumber Data**: 4 file CSV
- Buah-buahan: 2023 & 2024
- Sayur-sayuran: 2023 & 2024

---

### 🔄 **Tahapan Preprocessing yang Lengkap**:

| Tahap | Deskripsi | Output |
|-------|-----------|--------|
| **1. Data Loading** | Memuat file CSV dengan struktur header yang benar | 4 DataFrame mentah |
| **2. Data Cleaning** | Membersihkan missing values, inkonsistensi, duplikat | Data bersih & valid |
| **3. Data Transformation** | Konversi ke long format, normalisasi kolom | Data terstruktur |
| **4. Data Integration** | Menggabungkan semua dataset | Dataset terintegrasi |
| **5. Feature Engineering** | Membuat fitur tambahan (region, kategori, dll) | Data diperkaya |
| **6. Quality Check** | Validasi kualitas data | Laporan kualitas |
| **7. Export** | Simpan hasil ke CSV | 5 file clean data |

---

### ✅ **Data Siap Untuk**:
- 📈 **Exploratory Data Analysis (EDA)**
- 📊 **Data Visualization** 
- 📉 **Statistical Modeling**
- 🤖 **Machine Learning**

---

**Catatan**: Setiap tahapan dilengkapi dengan penjelasan detail pada markdown cell di bawahnya.

## 1. Import Libraries

**Tujuan**: Memuat semua library yang diperlukan untuk preprocessing data.

**Penjelasan**:
- `pandas`: Library utama untuk manipulasi dan analisis data tabular
- `numpy`: Library untuk operasi numerik dan array
- `warnings`: Untuk menyembunyikan peringatan yang tidak penting
- `os` & `Path`: Untuk manajemen file dan direktori

In [39]:
import pandas as pd
import numpy as np
import warnings
import os
from pathlib import Path
print("Libraries imported successfully!")

Libraries imported successfully!


## 2. Data Loading

**Tujuan**: Memuat data mentah dari file CSV dengan struktur header yang benar.

**Penjelasan**:
Dataset memiliki struktur khusus dengan beberapa baris header:
- **Baris 0**: Judul kolom pertama (Kabupaten/Kota)
- **Baris 1**: Deskripsi umum dataset
- **Baris 2**: Nama-nama produk (kolom yang akan digunakan)
- **Baris 3**: Tahun data
- **Baris 4+**: Data aktual

**Strategi**:
1. Membaca baris 2 untuk mendapatkan nama kolom (produk)
2. Melewati 4 baris pertama dan langsung membaca data aktual
3. Menggunakan nama kolom dari baris 2

**Dataset yang dimuat**:
- Buah-buahan: 2023 & 2024
- Sayur-sayuran: 2023 & 2024

**Output**: 4 DataFrame mentah dengan nama kolom yang benar (nama produk).

In [58]:
import pandas as pd

sayur = pd.read_csv(
    "./data/Rata-rata Pengeluaran Perkapita Seminggu  Menurut Kelompok Buah-Buahan Per Kabupaten_kota, 2023.csv"
)
buah = pd.read_csv(
    "./data/Rata-rata Pengeluaran Perkapita Seminggu  Menurut Kelompok Sayur-Sayuran Per Kabupaten_kota, 2023.csv"
)

sayur.head()
buah.head()

Kabupaten/Kota                                         Unnamed: 1  \
0            NaN  Rata-rata Pengeluaran Perkapita Seminggu Menur...   
1            NaN                                      SAYUR-SAYURAN   
2            NaN                                               2023   
3       Simeulue                                           12318.24   
4   Aceh Singkil                                           16723.57   

  Unnamed: 2 Unnamed: 3 Unnamed: 4           Unnamed: 5  Unnamed: 6  \
0        NaN        NaN        NaN                  NaN         NaN   
1      Bayam   Kangkung  Kol/kubis  Sawi putih (petsai)  Sawi hijau   
2       2023       2023       2023                 2023        2023   
3     354.36     782.14     202.72                71.08       80.88   
4     662.22     670.57     493.13               204.54      330.88   

  Unnamed: 7      Unnamed: 8               Unnamed: 9 Unnamed: 10 Unnamed: 11  \
0        NaN             NaN                      NaN         NaN         NaN   
1     Buncis  Kacang panjang  Tomat sayur, tomat ceri      Wortel    Mentimun   
2       2023            2023                     2023        2023        2023   
3      60.11          215.41                   272.89      174.06      260.51   
4     305.44          636.24                  1671.87      377.73      556.53   

                       Unnamed: 12 Unnamed: 13 Unnamed: 14  \
0                              NaN         NaN         NaN   
1  Daun ketela pohon/daun singkong      Terong       Tauge   
2                             2023        2023        2023   
3                           508.11      498.74      157.43   
4                           403.51      687.25      192.69   

                    Unnamed: 15                             Unnamed: 16  \
0                           NaN                                     NaN   
1  Labu, labu siam, labu parang  Bahan sayur sop/cap cay/ kimlo (paket)   
2                          2023                                    2023   
3                          76.6                                    5.28   
4                        106.91                                    3.45   

                      Unnamed: 17  Unnamed: 18  Unnamed: 19 Unnamed: 20  \
0                             NaN          NaN          NaN         NaN   
1  Bahan sayur asam/lodeh (paket)  Nangka muda  Pepaya muda     Jengkol   
2                            2023         2023         2023        2023   
3                           21.68        68.91         35.4       40.54   
4                            0.49        56.51        25.44      618.45   

    Unnamed: 21   Unnamed: 22    Unnamed: 23 Unnamed: 24 Unnamed: 25  \
0           NaN           NaN            NaN         NaN         NaN   
1  Bawang putih  Bawang merah  Bawang bombay  Cabe merah  Cabe hijau   
2          2023          2023           2023        2023        2023   
3       2265.97        595.58          22.92      3235.9      123.99   
4       2477.69        837.41         128.65      3866.9      440.46   

  Unnamed: 26            Unnamed: 27  
0         NaN                    NaN  
1  Cabe rawit  Sayur-sayuran lainnya  
2        2023                   2023  
3     2061.52                 125.51  
4      902.65                  65.97

In [40]:
data_dir = Path('data')
print("Loading datasets with proper headers...\n")

def load_csv_with_headers(filepath):
    df_temp = pd.read_csv(filepath, nrows=4, header=None)
    
    product_names = df_temp.iloc[2].tolist()
    
    df = pd.read_csv(filepath, skiprows=4, header=None, names=product_names)
    
    return df

# Buah-buahan
buah_2023_raw = load_csv_with_headers(data_dir / 'Rata-rata Pengeluaran Perkapita Seminggu  Menurut Kelompok Buah-Buahan Per Kabupaten_kota, 2023.csv')
buah_2024_raw = load_csv_with_headers(data_dir / 'Rata-rata Pengeluaran Perkapita Seminggu  Menurut Kelompok Buah-Buahan Per Kabupaten_kota, 2024.csv')

# Sayur-sayuran
sayur_2023_raw = load_csv_with_headers(data_dir / 'Rata-rata Pengeluaran Perkapita Seminggu  Menurut Kelompok Sayur-Sayuran Per Kabupaten_kota, 2023.csv')
sayur_2024_raw = load_csv_with_headers(data_dir / 'Rata-rata Pengeluaran Perkapita Seminggu  Menurut Kelompok Sayur-Sayuran Per Kabupaten_kota, 2024.csv')

print(f"Buah 2023: {buah_2023_raw.shape}")
print(f"  Columns: {list(buah_2023_raw.columns[:5])}...")
print(f"\nBuah 2024: {buah_2024_raw.shape}")
print(f"  Columns: {list(buah_2024_raw.columns[:5])}...")
print(f"\nSayur 2023: {sayur_2023_raw.shape}")
print(f"  Columns: {list(sayur_2023_raw.columns[:5])}...")
print(f"\nSayur 2024: {sayur_2024_raw.shape}")
print(f"  Columns: {list(sayur_2024_raw.columns[:5])}...")

print("\nData loaded successfully!")

Loading datasets with proper headers...

Buah 2023: (514, 17)
  Columns: [nan, 'BUAH-BUAHAN', 'Jeruk, jeruk bali', 'Mangga', 'Apel']...

Buah 2024: (514, 17)
  Columns: [nan, 'BUAH-BUAHAN', 'Jeruk, jeruk bali', 'Mangga', 'Apel']...

Sayur 2023: (514, 28)
  Columns: [nan, 'SAYUR-SAYURAN', 'Bayam', 'Kangkung', 'Kol/kubis']...

Sayur 2024: (514, 28)
  Columns: [nan, 'SAYUR-SAYURAN', 'Bayam', 'Kangkung', 'Kol/kubis']...

Data loaded successfully!


### 2.1 Custom CSV Loading Function

**Penjelasan Teknis**:

Fungsi `load_csv_with_headers()` dirancang khusus untuk menangani struktur CSV yang kompleks:

**Langkah-langkah**:
1. **Baca header** (4 baris pertama) untuk mendapatkan nama kolom dari baris ke-2
2. **Extract product names** dari baris index 2
3. **Load data actual** mulai dari baris ke-5 (index 4) dengan nama kolom yang sudah didapat

**Mengapa perlu custom function**:
- Fungsi `pd.read_csv()` biasa tidak bisa handle multi-level headers dengan baik
- Perlu kontrol manual untuk memilih baris mana yang jadi nama kolom
- Memastikan nama produk yang deskriptif (bukan '2023' atau 'nan')

**Parameter yang digunakan**:
- `nrows=4`: Hanya baca 4 baris pertama untuk ekstrak header
- `skiprows=4`: Lewati 4 baris pertama saat load data actual
- `header=None`: Jangan gunakan baris pertama sebagai header otomatis
- `names=product_names`: Gunakan nama kolom yang sudah kita ekstrak

In [41]:
print("BUAH 2023 - First 5 rows")
print(buah_2023_raw.head())
print("\nSAYUR 2023 - First 5 rows")
print(sayur_2023_raw.head())

BUAH 2023 - First 5 rows
             NaN  BUAH-BUAHAN Jeruk, jeruk bali  Mangga    Apel Rambutan  \
0       Simeulue      5287.59           1501.23   88.02  351.16    26.03   
1   Aceh Singkil      7377.18           1817.24  173.25  512.97    57.77   
2   Aceh Selatan      4589.20           1100.93  111.65  194.22    35.96   
3  Aceh Tenggara      5381.60            875.42  104.96   264.9   126.43   
4     Aceh Timur      5630.94           1670.13  240.13  142.95   114.39   

  Duku, langsat  Durian    Salak Pisang ambon  Pisang lainnya  Pepaya  \
0        118.96       -   699.85        19.36          898.80  396.75   
1        862.41   52.63  1098.24       151.88          571.33  273.94   
2        389.54    3.86   773.16       117.64          149.34  195.93   
3         280.8   275.9   909.91       198.38          434.09  334.67   
4        611.63  286.54  1097.17        68.22          248.52   298.8   

  Semangka Tomat buah Alpokat Jambu biji Buah-buahan lainnya  
0   263.62     3

### 2.1 Examine Raw Data Structure

**Tujuan**: Memahami struktur data mentah sebelum cleaning.

**Yang perlu diperhatikan**:
- Apakah header sudah benar?
- Apakah ada missing values?
- Bagaimana format nilai (numeric, string, dll)?
- Apakah ada anomali atau inkonsistensi?

**Analisis ini membantu** menentukan strategi cleaning yang tepat.

## 3. Data Cleaning

**Tujuan**: Membersihkan data dari inkonsistensi, missing values, dan error.

### 3.1 Cleaning Function - Remove Header Rows and Fix Structure

**Penjelasan**:
Fungsi `clean_dataset()` melakukan:
1. **Rename kolom pertama**: Mengubah nama kolom pertama (yang bernama `nan`) menjadi 'Kabupaten_Kota'
2. **Hapus baris kosong**: Menghapus baris yang tidak memiliki nama kabupaten/kota
3. **Tambah metadata**: Menambahkan kolom 'Tahun' dan 'Kategori' untuk identifikasi

**Mengapa diperlukan**:
- Data mentah memiliki struktur yang tidak konsisten
- Perlu identifikasi jelas untuk setiap baris (kabupaten, tahun, kategori)
- Memudahkan proses analisis selanjutnya

**Output**: DataFrame yang bersih dengan struktur kolom yang konsisten.

In [42]:
def clean_dataset(df, year, category):
    df_clean = df.copy()
    
    cols = list(df_clean.columns)
    cols[0] = 'Kabupaten_Kota'
    df_clean.columns = cols
    
    df_clean = df_clean[df_clean['Kabupaten_Kota'].notna()].copy()
    df_clean = df_clean[df_clean['Kabupaten_Kota'].astype(str).str.strip() != ''].copy()
    df_clean = df_clean[df_clean['Kabupaten_Kota'].astype(str).str.strip() != 'nan'].copy()
    
    df_clean['Tahun'] = year
    df_clean['Kategori'] = category
    
    print(f"Cleaned {category} {year}: {df_clean.shape}")
    print(f"  Columns: {list(df_clean.columns[:5])}...")
    
    return df_clean

print("Cleaning datasets...\n")
buah_2023 = clean_dataset(buah_2023_raw, 2023, 'Buah')
buah_2024 = clean_dataset(buah_2024_raw, 2024, 'Buah')
sayur_2023 = clean_dataset(sayur_2023_raw, 2023, 'Sayur')
sayur_2024 = clean_dataset(sayur_2024_raw, 2024, 'Sayur')

print("\nAll datasets cleaned!")

Cleaning datasets...

Cleaned Buah 2023: (514, 19)
  Columns: ['Kabupaten_Kota', 'BUAH-BUAHAN', 'Jeruk, jeruk bali', 'Mangga', 'Apel']...
Cleaned Buah 2024: (514, 19)
  Columns: ['Kabupaten_Kota', 'BUAH-BUAHAN', 'Jeruk, jeruk bali', 'Mangga', 'Apel']...
Cleaned Sayur 2023: (514, 30)
  Columns: ['Kabupaten_Kota', 'SAYUR-SAYURAN', 'Bayam', 'Kangkung', 'Kol/kubis']...
Cleaned Sayur 2024: (514, 30)
  Columns: ['Kabupaten_Kota', 'SAYUR-SAYURAN', 'Bayam', 'Kangkung', 'Kol/kubis']...

All datasets cleaned!


In [43]:
print("BUAH 2023 - Cleaned")
print(buah_2023.head())
print(f"\nColumns: {list(buah_2023.columns)}")
print(f"\nData types:\n{buah_2023.dtypes}")
print(f"\nShape: {buah_2023.shape}")
print(f"\nFirst column name: '{buah_2023.columns[0]}'")
print(f"\nSample values from first numeric column: {buah_2023.iloc[:5, 1].tolist()}")

BUAH 2023 - Cleaned
  Kabupaten_Kota  BUAH-BUAHAN Jeruk, jeruk bali  Mangga    Apel Rambutan  \
0       Simeulue      5287.59           1501.23   88.02  351.16    26.03   
1   Aceh Singkil      7377.18           1817.24  173.25  512.97    57.77   
2   Aceh Selatan      4589.20           1100.93  111.65  194.22    35.96   
3  Aceh Tenggara      5381.60            875.42  104.96   264.9   126.43   
4     Aceh Timur      5630.94           1670.13  240.13  142.95   114.39   

  Duku, langsat  Durian    Salak Pisang ambon  Pisang lainnya  Pepaya  \
0        118.96       -   699.85        19.36          898.80  396.75   
1        862.41   52.63  1098.24       151.88          571.33  273.94   
2        389.54    3.86   773.16       117.64          149.34  195.93   
3         280.8   275.9   909.91       198.38          434.09  334.67   
4        611.63  286.54  1097.17        68.22          248.52   298.8   

  Semangka Tomat buah Alpokat Jambu biji Buah-buahan lainnya  Tahun Kategori  
0   2

### 3.1.1 Validation: Examine Cleaned Data

**Tujuan**: Memvalidasi hasil cleaning dan memastikan struktur data sudah benar.

**Checklist Validasi**:
- ✅ Kolom pertama bernama 'Kabupaten_Kota'
- ✅ Nama kolom produk sudah benar (bukan '2023' atau 'nan')
- ✅ Kolom metadata 'Tahun' dan 'Kategori' sudah ditambahkan
- ✅ Data types sudah sesuai (akan dikonversi di step berikutnya)
- ✅ Tidak ada baris kosong atau invalid

**Interpretasi**:
Jika semua checklist terpenuhi, data siap untuk tahap konversi tipe data dan transformasi selanjutnya.

### 3.2 Handle Missing Values and Data Type Conversion

**Tujuan**: Menangani nilai yang hilang dan mengkonversi tipe data yang tepat.

**Masalah yang ditangani**:
1. **Simbol '-'**: Dalam data, tanda '-' menunjukkan tidak ada pengeluaran untuk produk tersebut
2. **String kosong**: Beberapa sel kosong perlu dikonversi ke nilai numerik
3. **Tipe data mixed**: Kolom yang seharusnya numerik masih bertipe object/string

**Strategi**:
- Ganti '-' dan string kosong dengan **0** (artinya: tidak ada pengeluaran)
- Konversi semua kolom produk menjadi **float64** (tipe numerik)
- Tetap pertahankan kolom Kabupaten_Kota, Tahun, Kategori sebagai string/integer

**Mengapa menggunakan 0 bukan NaN**:
- Nilai 0 lebih masuk akal secara bisnis (tidak ada pengeluaran)
- Memudahkan perhitungan statistik
- Tidak perlu handling khusus untuk missing values

**Output**: DataFrame dengan tipe data yang konsisten dan tidak ada missing values.

In [62]:
def handle_missing_and_convert(df):
    df_processed = df.copy()
    
    non_numeric_cols = ['Kabupaten_Kota', 'Tahun', 'Kategori']
    numeric_cols = [col for col in df_processed.columns if col not in non_numeric_cols]
    
    for col in numeric_cols:
        df_processed[col] = df_processed[col].replace('-', '0')
        df_processed[col] = df_processed[col].replace('', '0')
        
        df_processed[col] = pd.to_numeric(df_processed[col], errors='coerce')
    
    df_processed[numeric_cols] = df_processed[numeric_cols].fillna(0)
    
    df_processed['Kabupaten_Kota'] = df_processed['Kabupaten_Kota'].astype(str).str.strip()
    
    return df_processed

print("Processing missing values and converting data types...\n")
buah_2023_processed = handle_missing_and_convert(buah_2023)
buah_2024_processed = handle_missing_and_convert(buah_2024)
sayur_2023_processed = handle_missing_and_convert(sayur_2023)
sayur_2024_processed = handle_missing_and_convert(sayur_2024)

print("Missing values after processing:")
print(f"Buah 2023: {buah_2023_processed.isnull().sum().sum()}")
print(f"Buah 2024: {buah_2024_processed.isnull().sum().sum()}")
print(f"Sayur 2023: {sayur_2023_processed.isnull().sum().sum()}")
print(f"Sayur 2024: {sayur_2024_processed.isnull().sum().sum()}")
print("\nMissing values handled and data types converted!")

Processing missing values and converting data types...

Missing values after processing:
Buah 2023: 0
Buah 2024: 0
Sayur 2023: 0
Sayur 2024: 0

Missing values handled and data types converted!


### 3.3 Check for Duplicates

**Tujuan**: Memastikan tidak ada data duplikat berdasarkan Kabupaten/Kota.

**Penjelasan**:
- Setiap kabupaten/kota seharusnya hanya muncul **sekali** untuk setiap tahun
- Duplikat dapat menyebabkan bias dalam analisis
- Menggunakan `duplicated()` untuk deteksi dan `drop_duplicates()` untuk penghapusan

**Strategi**:
- Cek duplikat berdasarkan kolom 'Kabupaten_Kota'
- Jika ada duplikat, ambil data pertama (`keep='first'`)
- Laporkan jumlah duplikat yang ditemukan

**Output**: Dataset yang dijamin unik per kabupaten/kota per tahun.

In [60]:
print("Checking for duplicates...\n")
print(f"Buah 2023 duplicates: {buah_2023_processed.duplicated(subset=['Kabupaten_Kota']).sum()}")
print(f"Buah 2024 duplicates: {buah_2024_processed.duplicated(subset=['Kabupaten_Kota']).sum()}")
print(f"Sayur 2023 duplicates: {sayur_2023_processed.duplicated(subset=['Kabupaten_Kota']).sum()}")
print(f"Sayur 2024 duplicates: {sayur_2024_processed.duplicated(subset=['Kabupaten_Kota']).sum()}")

buah_2023_processed = buah_2023_processed.drop_duplicates(subset=['Kabupaten_Kota'], keep='first')
buah_2024_processed = buah_2024_processed.drop_duplicates(subset=['Kabupaten_Kota'], keep='first')
sayur_2023_processed = sayur_2023_processed.drop_duplicates(subset=['Kabupaten_Kota'], keep='first')
sayur_2024_processed = sayur_2024_processed.drop_duplicates(subset=['Kabupaten_Kota'], keep='first')

print("\nDuplicates removed!")

Checking for duplicates...

Buah 2023 duplicates: 0
Buah 2024 duplicates: 0
Sayur 2023 duplicates: 0
Sayur 2024 duplicates: 0

Duplicates removed!


### 3.4 Basic Statistics After Cleaning

**Tujuan**: Melihat ringkasan data setelah proses cleaning.

**Informasi yang ditampilkan**:
1. **Shape**: Jumlah baris dan kolom
2. **Unique count**: Berapa banyak kabupaten/kota unik
3. **Sample data**: Preview 3 baris pertama

**Mengapa penting**:
- Validasi bahwa cleaning berhasil
- Memahami skala dataset
- Identifikasi masalah potensial sebelum lanjut ke tahap berikutnya

**Expected output**:
- 514 kabupaten/kota (sesuai dengan jumlah kabupaten/kota di Indonesia)
- Semua kolom produk sudah bertipe float64
- Tidak ada nilai NaN

In [46]:
print("BUAH 2023 - Statistics")
print(f"Shape: {buah_2023_processed.shape}")
print(f"Kabupaten/Kota count: {buah_2023_processed['Kabupaten_Kota'].nunique()}")
print(f"\nSample data:")
print(buah_2023_processed.head(3))

print("\nSAYUR 2023 - Statistics")
print(f"Shape: {sayur_2023_processed.shape}")
print(f"Kabupaten/Kota count: {sayur_2023_processed['Kabupaten_Kota'].nunique()}")
print(f"\nSample data:")
print(sayur_2023_processed.head(3))

BUAH 2023 - Statistics
Shape: (514, 19)
Kabupaten/Kota count: 514

Sample data:
  Kabupaten_Kota  BUAH-BUAHAN  Jeruk, jeruk bali  Mangga    Apel  Rambutan  \
0       Simeulue      5287.59            1501.23   88.02  351.16     26.03   
1   Aceh Singkil      7377.18            1817.24  173.25  512.97     57.77   
2   Aceh Selatan      4589.20            1100.93  111.65  194.22     35.96   

   Duku, langsat  Durian    Salak  Pisang ambon  Pisang lainnya  Pepaya  \
0         118.96    0.00   699.85         19.36          898.80  396.75   
1         862.41   52.63  1098.24        151.88          571.33  273.94   
2         389.54    3.86   773.16        117.64          149.34  195.93   

   Semangka  Tomat buah  Alpokat  Jambu biji  Buah-buahan lainnya  Tahun  \
0    263.62      388.53   126.78       38.27               370.22   2023   
1   1061.05        8.49   128.52       25.65               581.79   2023   
2    201.34     1027.09   149.77        4.42               134.34   2023   

 

## 4. Data Transformation

**Tujuan**: Mengubah format dan struktur data agar lebih sesuai untuk analisis.

### 4.1 Normalize Column Names

**Penjelasan**:
- **Membersihkan nama kolom**: Menghapus spasi ekstra
- **Mempertahankan readability**: Nama produk tetap jelas (contoh: "Jeruk, jeruk bali" bukan "jeruk_jeruk_bali")
- **Konsistensi**: Semua nama kolom mengikuti format yang sama

**Mengapa penting**:
- Nama kolom yang konsisten memudahkan akses data
- Menghindari error karena spasi atau karakter khusus
- Mempermudah dokumentasi dan maintenance code

**Catatan**: Kita mempertahankan nama produk yang deskriptif agar mudah dipahami oleh stakeholder non-teknis.

In [47]:
def normalize_column_names(df):
    df_norm = df.copy()
    
    col_mapping = {}
    for col in df_norm.columns:
        if col not in ['Kabupaten_Kota', 'Tahun', 'Kategori']:
            new_col = col.strip()
            col_mapping[col] = new_col
    
    df_norm.rename(columns=col_mapping, inplace=True)
    
    return df_norm

buah_2023_norm = normalize_column_names(buah_2023_processed)
buah_2024_norm = normalize_column_names(buah_2024_processed)
sayur_2023_norm = normalize_column_names(sayur_2023_processed)
sayur_2024_norm = normalize_column_names(sayur_2024_processed)

print("Column names normalized!")
print(f"\nBuah columns: {list(buah_2023_norm.columns[:5])}...")
print(f"Sayur columns: {list(sayur_2023_norm.columns[:5])}...")

Column names normalized!

Buah columns: ['Kabupaten_Kota', 'BUAH-BUAHAN', 'Jeruk, jeruk bali', 'Mangga', 'Apel']...
Sayur columns: ['Kabupaten_Kota', 'SAYUR-SAYURAN', 'Bayam', 'Kangkung', 'Kol/kubis']...


### 4.2 Create Long Format for Analysis (Melt Data)

**Tujuan**: Mengubah data dari format wide (lebar) menjadi long (panjang).

**Format Wide vs Long**:

**Wide Format** (sebelum):
```
Kabupaten | Apel | Mangga | Jeruk
Jakarta   | 100  | 200    | 150
Bandung   | 120  | 180    | 160
```

**Long Format** (sesudah):
```
Kabupaten | Produk | Pengeluaran
Jakarta   | Apel   | 100
Jakarta   | Mangga | 200
Jakarta   | Jeruk  | 150
Bandung   | Apel   | 120
...
```

**Keuntungan Long Format**:
1. ✅ Lebih fleksibel untuk filtering dan grouping
2. ✅ Ideal untuk visualisasi dengan library seperti Seaborn, Plotly
3. ✅ Sesuai dengan prinsip "tidy data"
4. ✅ Memudahkan analisis time-series
5. ✅ Lebih efisien untuk machine learning

**Kolom hasil**:
- `Kabupaten_Kota`: Nama wilayah
- `Tahun`: 2023 atau 2024
- `Kategori`: Buah atau Sayur
- `Produk`: Nama produk spesifik
- `Pengeluaran`: Nilai pengeluaran (Rupiah/Kapita/Minggu)
- `Is_Total`: Flag untuk kolom total kategori

**Output**: Dataset dalam format panjang yang siap untuk analisis mendalam.

In [48]:
def transform_to_long_format(df):
    id_vars = ['Kabupaten_Kota', 'Tahun', 'Kategori']
    
    product_cols = [col for col in df.columns if col not in id_vars]
    total_col = product_cols[0]
    
    df_long = df.melt(
        id_vars=id_vars,
        value_vars=product_cols,
        var_name='Produk',
        value_name='Pengeluaran'
    )
    
    df_long['Is_Total'] = df_long['Produk'] == total_col
    
    return df_long

print("Transforming to long format...\n")
buah_2023_long = transform_to_long_format(buah_2023_norm)
buah_2024_long = transform_to_long_format(buah_2024_norm)
sayur_2023_long = transform_to_long_format(sayur_2023_norm)
sayur_2024_long = transform_to_long_format(sayur_2024_norm)

print(f"Buah 2023 long: {buah_2023_long.shape}")
print(f"Buah 2024 long: {buah_2024_long.shape}")
print(f"Sayur 2023 long: {sayur_2023_long.shape}")
print(f"Sayur 2024 long: {sayur_2024_long.shape}")

print("\nData transformed to long format!")
print("\nSample long format data:")
print(buah_2023_long.head(10))

Transforming to long format...

Buah 2023 long: (8224, 6)
Buah 2024 long: (8224, 6)
Sayur 2023 long: (13878, 6)
Sayur 2024 long: (13878, 6)

Data transformed to long format!

Sample long format data:
  Kabupaten_Kota  Tahun Kategori       Produk  Pengeluaran  Is_Total
0       Simeulue   2023     Buah  BUAH-BUAHAN      5287.59      True
1   Aceh Singkil   2023     Buah  BUAH-BUAHAN      7377.18      True
2   Aceh Selatan   2023     Buah  BUAH-BUAHAN      4589.20      True
3  Aceh Tenggara   2023     Buah  BUAH-BUAHAN      5381.60      True
4     Aceh Timur   2023     Buah  BUAH-BUAHAN      5630.94      True
5    Aceh Tengah   2023     Buah  BUAH-BUAHAN     10500.46      True
6     Aceh Barat   2023     Buah  BUAH-BUAHAN      9451.44      True
7     Aceh Besar   2023     Buah  BUAH-BUAHAN      5480.57      True
8          Pidie   2023     Buah  BUAH-BUAHAN      5716.10      True
9        Bireuen   2023     Buah  BUAH-BUAHAN      7543.48      True


## 5. Data Integration

**Tujuan**: Menggabungkan semua dataset menjadi satu dataset terintegrasi.

### 5.1 Combine All Datasets

**Penjelasan**:
Menggabungkan 4 dataset menjadi 1 dataset unified:
- Buah 2023 (8,224 baris)
- Buah 2024 (8,224 baris)
- Sayur 2023 (13,878 baris)
- Sayur 2024 (13,878 baris)

**Hasil**: Total 44,204 baris data terintegrasi

**Keuntungan integrasi**:
1. ✅ **Analisis komparatif**: Membandingkan tahun 2023 vs 2024
2. ✅ **Analisis cross-category**: Membandingkan Buah vs Sayur
3. ✅ **Analisis holistik**: Melihat pola konsumsi secara keseluruhan
4. ✅ **Single source of truth**: Satu dataset untuk semua analisis

**Metadata yang tersimpan**:
- **Tahun**: Untuk analisis time-series
- **Kategori**: Untuk segmentasi Buah vs Sayur
- **Produk**: Untuk analisis detail per item

**Output**: Dataset terintegrasi dalam format long dengan 6 kolom utama.

In [49]:
print("Integrating all datasets...\n")

df_integrated_long = pd.concat([
    buah_2023_long,
    buah_2024_long,
    sayur_2023_long,
    sayur_2024_long
], ignore_index=True)

print(f"Integrated long format dataset: {df_integrated_long.shape}")
print(f"\nUnique years: {sorted(df_integrated_long['Tahun'].unique())}")
print(f"Unique categories: {df_integrated_long['Kategori'].unique()}")
print(f"Unique Kabupaten/Kota: {df_integrated_long['Kabupaten_Kota'].nunique()}")
print(f"Unique products: {df_integrated_long['Produk'].nunique()}")

print("\nSample Integrated Data")
print(df_integrated_long.head(10))

Integrating all datasets...

Integrated long format dataset: (44204, 6)

Unique years: [np.int64(2023), np.int64(2024)]
Unique categories: ['Buah' 'Sayur']
Unique Kabupaten/Kota: 514
Unique products: 43

Sample Integrated Data
  Kabupaten_Kota  Tahun Kategori       Produk  Pengeluaran  Is_Total
0       Simeulue   2023     Buah  BUAH-BUAHAN      5287.59      True
1   Aceh Singkil   2023     Buah  BUAH-BUAHAN      7377.18      True
2   Aceh Selatan   2023     Buah  BUAH-BUAHAN      4589.20      True
3  Aceh Tenggara   2023     Buah  BUAH-BUAHAN      5381.60      True
4     Aceh Timur   2023     Buah  BUAH-BUAHAN      5630.94      True
5    Aceh Tengah   2023     Buah  BUAH-BUAHAN     10500.46      True
6     Aceh Barat   2023     Buah  BUAH-BUAHAN      9451.44      True
7     Aceh Besar   2023     Buah  BUAH-BUAHAN      5480.57      True
8          Pidie   2023     Buah  BUAH-BUAHAN      5716.10      True
9        Bireuen   2023     Buah  BUAH-BUAHAN      7543.48      True


### 5.2 Create Wide Format Integrated Dataset (for specific analyses)

**Tujuan**: Membuat dataset format wide yang menggabungkan Buah dan Sayur.

**Penjelasan**:
Meskipun format long lebih fleksibel, format wide tetap berguna untuk:
- Analisis korelasi antar produk
- Regresi dengan multiple features
- Pivot tables dan crosstabs
- Ekspor ke Excel untuk stakeholder

**Proses**:
1. Gabungkan data Buah 2023 & 2024
2. Gabungkan data Sayur 2023 & 2024
3. Tambahkan prefix pada nama kolom (Buah_ dan Sayur_)
4. Merge berdasarkan Kabupaten_Kota dan Tahun

**Hasil**:
- 1,028 baris (514 kabupaten × 2 tahun)
- 47 kolom (metadata + semua produk Buah & Sayur)

**Use case**:
- Analisis: "Apakah kabupaten yang banyak beli buah juga banyak beli sayur?"
- Model prediksi dengan semua fitur sekaligus
- Dashboard dengan multiple metrics

**Output**: Dataset wide yang comprehensive dengan semua fitur.

In [50]:
df_buah_combined = pd.concat([buah_2023_norm, buah_2024_norm], ignore_index=True)
df_sayur_combined = pd.concat([sayur_2023_norm, sayur_2024_norm], ignore_index=True)

print("Wide format combined datasets:")
print(f"Buah combined: {df_buah_combined.shape}")
print(f"Sayur combined: {df_sayur_combined.shape}")

df_buah_combined_renamed = df_buah_combined.copy()
df_sayur_combined_renamed = df_sayur_combined.copy()

for col in df_buah_combined_renamed.columns:
    if col not in ['Kabupaten_Kota', 'Tahun', 'Kategori']:
        df_buah_combined_renamed.rename(columns={col: f'Buah_{col}'}, inplace=True)

for col in df_sayur_combined_renamed.columns:
    if col not in ['Kabupaten_Kota', 'Tahun', 'Kategori']:
        df_sayur_combined_renamed.rename(columns={col: f'Sayur_{col}'}, inplace=True)

df_integrated_wide = df_buah_combined_renamed.merge(
    df_sayur_combined_renamed,
    on=['Kabupaten_Kota', 'Tahun'],
    how='outer',
    suffixes=('_Buah', '_Sayur')
)

print(f"\nFully integrated wide format: {df_integrated_wide.shape}")
print(f"\nColumns: {len(df_integrated_wide.columns)}")

Wide format combined datasets:
Buah combined: (1028, 19)
Sayur combined: (1028, 30)

Fully integrated wide format: (1028, 47)

Columns: 47


## 6. Feature Engineering & Additional Transformations

**Tujuan**: Membuat fitur-fitur baru (derived features) untuk memperkaya analisis.

**Penjelasan**:
Feature engineering adalah proses membuat variabel baru dari data yang ada untuk meningkatkan kualitas analisis dan model.

**Fitur yang dibuat**:

1. **Region** (Klasifikasi Wilayah)
   - Identifikasi wilayah berdasarkan nama kabupaten
   - Contoh: Kabupaten dengan "Aceh" → Region "Aceh"
   - Berguna untuk analisis regional

2. **Kategori_Pengeluaran** (Kategorisasi Nilai)
   - Mengubah nilai numerik menjadi kategori deskriptif
   - Rentang: Tidak Ada, Sangat Rendah, Rendah, Sedang, Tinggi, Sangat Tinggi
   - Memudahkan segmentasi dan visualisasi

3. **Pengeluaran_Bulanan**
   - Konversi dari mingguan ke bulanan (× 4)
   - Lebih intuitif untuk perencanaan budget

4. **Pengeluaran_Tahunan**
   - Konversi dari mingguan ke tahunan (× 52)
   - Untuk analisis ekonomi makro dan proyeksi

**Mengapa Feature Engineering Penting**:
✅ Meningkatkan interpretability data
✅ Memudahkan pembuatan visualisasi
✅ Memperkaya context untuk machine learning
✅ Memenuhi kebutuhan analisis bisnis

**Output**: Dataset yang diperkaya dengan 4 fitur tambahan yang bermakna.

In [51]:
df_final_long = df_integrated_long.copy()

print("Adding derived features...\n")

df_final_long['Region'] = df_final_long['Kabupaten_Kota'].apply(
    lambda x: 'Aceh' if 'Aceh' in str(x) else 'Lainnya'
)

def categorize_expenditure(value):
    if value == 0:
        return 'Tidak Ada'
    elif value < 100:
        return 'Sangat Rendah'
    elif value < 500:
        return 'Rendah'
    elif value < 1000:
        return 'Sedang'
    elif value < 2000:
        return 'Tinggi'
    else:
        return 'Sangat Tinggi'

df_final_long['Kategori_Pengeluaran'] = df_final_long['Pengeluaran'].apply(categorize_expenditure)

df_final_long['Pengeluaran_Bulanan'] = df_final_long['Pengeluaran'] * 4

df_final_long['Pengeluaran_Tahunan'] = df_final_long['Pengeluaran'] * 52

print("Derived features added!")
print(f"\nNew columns: {list(df_final_long.columns)}")
print(f"\nCategories of expenditure:")
print(df_final_long['Kategori_Pengeluaran'].value_counts())

Adding derived features...

Derived features added!

New columns: ['Kabupaten_Kota', 'Tahun', 'Kategori', 'Produk', 'Pengeluaran', 'Is_Total', 'Region', 'Kategori_Pengeluaran', 'Pengeluaran_Bulanan', 'Pengeluaran_Tahunan']

Categories of expenditure:
Kategori_Pengeluaran
Rendah           18289
Sangat Rendah     8786
Sedang            7582
Tinggi            4301
Sangat Tinggi     4265
Tidak Ada          981
Name: count, dtype: int64


## 7. Data Quality Report

**Tujuan**: Melakukan validasi komprehensif terhadap kualitas data hasil preprocessing.

**Aspek yang divalidasi**:

### 1. **Dataset Overview**
- Jumlah total records
- Jumlah fitur/kolom
- Memory usage (efisiensi penyimpanan)

### 2. **Data Completeness**
- Cek missing values
- Persentase kelengkapan data
- Target: 100% completeness

### 3. **Data Distribution**
- Range tahun yang tersedia
- Kategori produk (Buah/Sayur)
- Jumlah wilayah dan produk unik

### 4. **Expenditure Statistics**
- Mean, median, std deviation
- Min dan max values
- Distribusi pengeluaran

### 5. **Zero Expenditure Analysis**
- Berapa banyak record dengan nilai 0
- Apakah wajar atau anomali

### 6. **Data Types Validation**
- Memastikan tipe data sesuai
- Numeric untuk nilai, object untuk kategori

**Interpretasi**:
- **100% completeness** = Data siap pakai
- **Zero expenditure <5%** = Wajar (beberapa daerah memang tidak konsumsi produk tertentu)
- **Consistent data types** = Tidak akan ada error saat analisis

**Output**: Laporan lengkap kualitas data yang memberikan confidence untuk analisis lanjutan.

In [52]:
print("="*60)
print("DATA QUALITY REPORT")
print("="*60)

print("\n1. DATASET OVERVIEW")
print(f"   Total records: {len(df_final_long):,}")
print(f"   Total features: {len(df_final_long.columns)}")
print(f"   Memory usage: {df_final_long.memory_usage(deep=True).sum() / 1024**2:.2f} MB")

print("\n2. DATA COMPLETENESS")
print(f"   Missing values: {df_final_long.isnull().sum().sum()}")
print(f"   Completeness: {(1 - df_final_long.isnull().sum().sum() / df_final_long.size) * 100:.2f}%")

print("\n3. DATA DISTRIBUTION")
print(f"   Years: {sorted(df_final_long['Tahun'].unique())}")
print(f"   Categories: {df_final_long['Kategori'].unique()}")
print(f"   Kabupaten/Kota: {df_final_long['Kabupaten_Kota'].nunique()}")
print(f"   Products: {df_final_long['Produk'].nunique()}")

print("\n4. EXPENDITURE STATISTICS")
print(df_final_long['Pengeluaran'].describe())

print("\n5. ZERO EXPENDITURE")
zero_expenditure = (df_final_long['Pengeluaran'] == 0).sum()
print(f"   Records with zero expenditure: {zero_expenditure:,} ({zero_expenditure/len(df_final_long)*100:.2f}%)")

print("\n6. DATA TYPES")
print(df_final_long.dtypes)

print("\n" + "="*60)
print("DATA QUALITY CHECK COMPLETE!")
print("="*60)

DATA QUALITY REPORT

1. DATASET OVERVIEW
   Total records: 44,204
   Total features: 10
   Memory usage: 15.21 MB

2. DATA COMPLETENESS
   Missing values: 0
   Completeness: 100.00%

3. DATA DISTRIBUTION
   Years: [np.int64(2023), np.int64(2024)]
   Categories: ['Buah' 'Sayur']
   Kabupaten/Kota: 514
   Products: 43

4. EXPENDITURE STATISTICS
count    44204.000000
mean      1059.547397
std       2733.244628
min          0.000000
25%        117.397500
50%        322.000000
75%        759.707500
max      81168.000000
Name: Pengeluaran, dtype: float64

5. ZERO EXPENDITURE
   Records with zero expenditure: 981 (2.22%)

6. DATA TYPES
Kabupaten_Kota           object
Tahun                     int64
Kategori                 object
Produk                   object
Pengeluaran             float64
Is_Total                   bool
Region                   object
Kategori_Pengeluaran     object
Pengeluaran_Bulanan     float64
Pengeluaran_Tahunan     float64
dtype: object

DATA QUALITY CHECK COMPLETE!

## 8. Export Cleaned Data

**Tujuan**: Menyimpan data yang sudah diproses ke file CSV untuk penggunaan lebih lanjut.

**File yang dihasilkan**:

### 1. **data_integrated_long.csv** ⭐ UTAMA
- Format: Long format (44,204 baris × 10 kolom)
- **Recommended untuk**: EDA, Visualisasi, Machine Learning
- Paling fleksibel dan mudah digunakan

### 2. **data_integrated_wide.csv**
- Format: Wide format (1,028 baris × 47 kolom)
- **Recommended untuk**: Analisis korelasi, Regresi, Dashboard Excel

### 3. **data_buah_combined.csv**
- Data khusus buah-buahan (semua tahun)
- **Use case**: Analisis spesifik kategori buah

### 4. **data_sayur_combined.csv**
- Data khusus sayur-sayuran (semua tahun)
- **Use case**: Analisis spesifik kategori sayur

### 5. **summary_statistics.csv**
- Statistik ringkasan per produk per tahun
- **Use case**: Quick reference, Reporting

**Lokasi**: `data/cleaned/`

**Keuntungan Export**:
✅ Reproducibility: Data bisa digunakan ulang tanpa re-processing
✅ Sharing: Mudah dibagikan ke tim lain
✅ Backup: Checkpoint jika perlu rollback
✅ Performance: Loading CSV lebih cepat dari re-processing

**Format CSV dipilih karena**:
- Universal dan kompatibel dengan semua tools
- Human-readable (bisa dibuka di Excel)
- Lightweight dan efisien

In [53]:
output_dir = Path('data/cleaned')
output_dir.mkdir(parents=True, exist_ok=True)

print("Exporting cleaned datasets...\n")

df_final_long.to_csv(output_dir / 'data_integrated_long.csv', index=False)
print(f"Exported: data_integrated_long.csv ({df_final_long.shape})")

df_integrated_wide.to_csv(output_dir / 'data_integrated_wide.csv', index=False)
print(f"Exported: data_integrated_wide.csv ({df_integrated_wide.shape})")

df_buah_combined.to_csv(output_dir / 'data_buah_combined.csv', index=False)
print(f"Exported: data_buah_combined.csv ({df_buah_combined.shape})")

df_sayur_combined.to_csv(output_dir / 'data_sayur_combined.csv', index=False)
print(f"Exported: data_sayur_combined.csv ({df_sayur_combined.shape})")

summary_stats = df_final_long.groupby(['Kategori', 'Tahun', 'Produk']).agg({
    'Pengeluaran': ['mean', 'median', 'std', 'min', 'max', 'count']
}).round(2)
summary_stats.to_csv(output_dir / 'summary_statistics.csv')
print(f"Exported: summary_statistics.csv")

print(f"\nAll files exported to: {output_dir.absolute()}")
print("\n" + "="*60)
print("PREPROCESSING COMPLETE!")
print("="*60)
print("\nData is ready for:")
print("  Exploratory Data Analysis (EDA)")
print("  Statistical Modeling")
print("  Machine Learning")
print("  Visualization")

Exporting cleaned datasets...

Exported: data_integrated_long.csv ((44204, 10))
Exported: data_integrated_wide.csv ((1028, 47))
Exported: data_buah_combined.csv ((1028, 19))
Exported: data_sayur_combined.csv ((1028, 30))
Exported: summary_statistics.csv

All files exported to: d:\Perkuliahan\2025-2026\Analisa Big Data\Project Real\data\cleaned

PREPROCESSING COMPLETE!

Data is ready for:
  Exploratory Data Analysis (EDA)
  Statistical Modeling
  Machine Learning
  Visualization
Exported: data_integrated_long.csv ((44204, 10))
Exported: data_integrated_wide.csv ((1028, 47))
Exported: data_buah_combined.csv ((1028, 19))
Exported: data_sayur_combined.csv ((1028, 30))
Exported: summary_statistics.csv

All files exported to: d:\Perkuliahan\2025-2026\Analisa Big Data\Project Real\data\cleaned

PREPROCESSING COMPLETE!

Data is ready for:
  Exploratory Data Analysis (EDA)
  Statistical Modeling
  Machine Learning
  Visualization


## 9. Quick Data Preview for EDA

**Tujuan**: Menampilkan preview dan insights awal dari data yang sudah bersih.

**Informasi yang ditampilkan**:

### 1. **Final Dataset Overview**
- Struktur data final
- Kolom-kolom yang tersedia
- Sample data untuk validasi

### 2. **Aggregated Statistics**
- Rata-rata pengeluaran per kategori dan tahun
- Total pengeluaran untuk analisis ekonomi
- Jumlah data points per segment

### 3. **Top Products Analysis**
- 10 produk dengan pengeluaran tertinggi
- Insight tentang preferensi konsumsi
- Identifikasi produk premium vs basic

### 4. **Regional Comparison**
- Perbandingan pengeluaran antar region
- Pola konsumsi geografis
- Potensi untuk analisis disparitas regional

**Manfaat Preview**:
✅ Quick validation bahwa data sudah benar
✅ Early insights sebelum deep dive analysis
✅ Membantu merencanakan analisis lanjutan
✅ Mengidentifikasi pola menarik untuk dieksplorasi

**Next Steps setelah preview**:
1. Buat visualisasi (charts, graphs)
2. Statistical testing (hypothesis testing)
3. Time series analysis
4. Predictive modeling

In [54]:
print("="*60)
print("FINAL CLEANED DATASET - LONG FORMAT")
print("="*60)
print(f"\nShape: {df_final_long.shape}")
print(f"\nColumns: {list(df_final_long.columns)}")
print("\nFirst 10 rows:")
print(df_final_long.head(10))

print("\n" + "="*60)
print("\nSample by Category:")
print(df_final_long.groupby(['Kategori', 'Tahun']).agg({
    'Pengeluaran': ['mean', 'sum', 'count']
}).round(2))

print("\n" + "="*60)
print("\nTop 10 Products by Average Expenditure:")
top_products = df_final_long.groupby('Produk')['Pengeluaran'].mean().sort_values(ascending=False).head(10)
print(top_products.round(2))

print("\n" + "="*60)
print("\nData by Region:")
print(df_final_long.groupby(['Region', 'Kategori', 'Tahun'])['Pengeluaran'].mean().round(2))

FINAL CLEANED DATASET - LONG FORMAT

Shape: (44204, 10)

Columns: ['Kabupaten_Kota', 'Tahun', 'Kategori', 'Produk', 'Pengeluaran', 'Is_Total', 'Region', 'Kategori_Pengeluaran', 'Pengeluaran_Bulanan', 'Pengeluaran_Tahunan']

First 10 rows:
  Kabupaten_Kota  Tahun Kategori       Produk  Pengeluaran  Is_Total   Region  \
0       Simeulue   2023     Buah  BUAH-BUAHAN      5287.59      True  Lainnya   
1   Aceh Singkil   2023     Buah  BUAH-BUAHAN      7377.18      True     Aceh   
2   Aceh Selatan   2023     Buah  BUAH-BUAHAN      4589.20      True     Aceh   
3  Aceh Tenggara   2023     Buah  BUAH-BUAHAN      5381.60      True     Aceh   
4     Aceh Timur   2023     Buah  BUAH-BUAHAN      5630.94      True     Aceh   
5    Aceh Tengah   2023     Buah  BUAH-BUAHAN     10500.46      True     Aceh   
6     Aceh Barat   2023     Buah  BUAH-BUAHAN      9451.44      True     Aceh   
7     Aceh Besar   2023     Buah  BUAH-BUAHAN      5480.57      True     Aceh   
8          Pidie   2023     Buah

## 📋 Summary - Preprocessing Complete!

---

### ✅ **Tahapan yang Telah Selesai**:

#### **1. Data Loading** ✓
- ✅ Loaded 4 CSV files dengan custom function untuk handle multi-level headers
- ✅ Total: 514 kabupaten/kota × 2 tahun × 2 kategori
- ✅ Produk: 16 jenis buah + 27 jenis sayur

#### **2. Data Cleaning** ✓
- ✅ Removed header rows yang tidak perlu
- ✅ Handled missing values (replaced '-' dengan 0)
- ✅ Converted semua kolom numerik ke float64
- ✅ Removed 0 duplicates
- ✅ Cleaned nama kabupaten/kota

#### **3. Data Transformation** ✓
- ✅ Normalized column names
- ✅ Created **long format** (44,204 rows × 10 columns) - RECOMMENDED
- ✅ Created **wide format** (1,028 rows × 47 columns) - for specific use cases

#### **4. Data Integration** ✓
- ✅ Combined Buah 2023 + Buah 2024 + Sayur 2023 + Sayur 2024
- ✅ Maintained metadata: Tahun, Kategori, Kabupaten_Kota
- ✅ Created unified dataset untuk analisis comprehensive

#### **5. Feature Engineering** ✓
- ✅ **Region**: Klasifikasi wilayah geografis
- ✅ **Kategori_Pengeluaran**: 6 level (Tidak Ada → Sangat Tinggi)
- ✅ **Pengeluaran_Bulanan**: Konversi mingguan ke bulanan (× 4)
- ✅ **Pengeluaran_Tahunan**: Konversi mingguan ke tahunan (× 52)

#### **6. Data Quality Check** ✓
- ✅ **Completeness**: 100% (no missing values)
- ✅ **Total records**: 44,204
- ✅ **Memory usage**: 15.21 MB (efficient)
- ✅ **Zero expenditure**: 2.22% (acceptable)
- ✅ **Data types**: Consistent and correct

#### **7. Export** ✓
- ✅ `data_integrated_long.csv` - Main dataset (RECOMMENDED)
- ✅ `data_integrated_wide.csv` - Wide format
- ✅ `data_buah_combined.csv` - Fruits only
- ✅ `data_sayur_combined.csv` - Vegetables only
- ✅ `summary_statistics.csv` - Quick statistics

---

### 📊 **Quick Insights**:

| Metric | Value |
|--------|-------|
| **Rata-rata pengeluaran** | Rp 1,059.55 / kapita / minggu |
| **Produk termahal** | SAYUR-SAYURAN (Rp 14,553.52) |
| **Produk ke-2** | BUAH-BUAHAN (Rp 8,226.74) |
| **Bumbu termahal** | Bawang putih (Rp 2,072.85) |
| **Wilayah** | 514 Kabupaten/Kota |
| **Produk unik** | 43 items |
| **Periode** | 2023 - 2024 |

---

### 🎯 **Data Siap Untuk**:

| Analisis | Tools | File yang Digunakan |
|----------|-------|---------------------|
| **EDA** | Pandas, Numpy | `data_integrated_long.csv` |
| **Visualization** | Matplotlib, Seaborn, Plotly | `data_integrated_long.csv` |
| **Statistical Modeling** | Scipy, Statsmodels | `data_integrated_wide.csv` |
| **Machine Learning** | Scikit-learn, TensorFlow | `data_integrated_long.csv` |
| **Dashboard/BI** | Tableau, PowerBI, Excel | `data_integrated_wide.csv` |

---

### 📁 **Output Files Location**:
```
data/cleaned/
├── data_integrated_long.csv    ⭐ UTAMA
├── data_integrated_wide.csv
├── data_buah_combined.csv
├── data_sayur_combined.csv
└── summary_statistics.csv
```

---

### 🚀 **Next Steps - Recommended Analysis**:

1. **Time Series Analysis**
   - Trend pengeluaran 2023 vs 2024
   - Seasonal patterns
   - Growth rate per produk

2. **Comparative Analysis**
   - Buah vs Sayur consumption
   - Regional disparities
   - Urban vs Rural (if data available)

3. **Clustering**
   - Kabupaten grouping by consumption pattern
   - Product bundling analysis

4. **Predictive Modeling**
   - Forecast 2025 expenditure
   - Demand prediction per product

---

**✨ Preprocessing Complete - Data Ready for Analysis! ✨**

In [55]:
print("="*60)
print("QUICK ACCESS TO CLEANED DATASETS")
print("="*60)

print("\n1. MAIN DATASET (Long Format) - Recommended for EDA & Modeling")
print(f"   Variable: df_final_long")
print(f"   Shape: {df_final_long.shape}")
print(f"   File: data/cleaned/data_integrated_long.csv")

print("\n2. WIDE FORMAT DATASET - For specific analyses")
print(f"   Variable: df_integrated_wide")
print(f"   Shape: {df_integrated_wide.shape}")
print(f"   File: data/cleaned/data_integrated_wide.csv")

print("\n3. INDIVIDUAL CATEGORY DATASETS")
print(f"   Buah: df_buah_combined ({df_buah_combined.shape})")
print(f"   File: data/cleaned/data_buah_combined.csv")
print(f"   Sayur: df_sayur_combined ({df_sayur_combined.shape})")
print(f"   File: data/cleaned/data_sayur_combined.csv")

print("\n" + "="*60)
print("You can now proceed with:")
print("  1. Exploratory Data Analysis (EDA)")
print("  2. Data Visualization")
print("  3. Statistical Analysis")
print("  4. Machine Learning Modeling")
print("="*60)

QUICK ACCESS TO CLEANED DATASETS

1. MAIN DATASET (Long Format) - Recommended for EDA & Modeling
   Variable: df_final_long
   Shape: (44204, 10)
   File: data/cleaned/data_integrated_long.csv

2. WIDE FORMAT DATASET - For specific analyses
   Variable: df_integrated_wide
   Shape: (1028, 47)
   File: data/cleaned/data_integrated_wide.csv

3. INDIVIDUAL CATEGORY DATASETS
   Buah: df_buah_combined ((1028, 19))
   File: data/cleaned/data_buah_combined.csv
   Sayur: df_sayur_combined ((1028, 30))
   File: data/cleaned/data_sayur_combined.csv

You can now proceed with:
  1. Exploratory Data Analysis (EDA)
  2. Data Visualization
  3. Statistical Analysis
  4. Machine Learning Modeling


---

## 📚 **Penjelasan Lengkap Setiap Tahapan**

Notebook ini telah dilengkapi dengan **penjelasan detail** untuk setiap tahapan preprocessing:

### ✅ **Yang Sudah Dijelaskan**:

1. **Import Libraries** - Mengapa library tertentu dipilih dan kegunaannya
2. **Data Loading** - Teknik custom loading untuk handle multi-level headers
3. **Data Cleaning** - Strategi menangani missing values, duplikat, dan tipe data
4. **Data Transformation** - Perbedaan wide vs long format dan keuntungannya
5. **Data Integration** - Proses menggabungkan multiple datasets
6. **Feature Engineering** - Pembuatan fitur turunan untuk analisis lebih kaya
7. **Quality Report** - Validasi komprehensif kualitas data
8. **Export** - Penjelasan format file dan use case masing-masing

### 📖 **Cara Membaca Notebook**:

- **Markdown cells** (warna biru): Penjelasan konsep dan metodologi
- **Code cells** (warna abu): Implementasi actual dengan comments
- **Output cells**: Hasil eksekusi untuk validasi

### 🎯 **Rekomendasi Next Steps**:

1. **Exploratory Data Analysis (EDA)**:
   - Gunakan file: `data_integrated_long.csv`
   - Analisis: Distribusi, trend, pattern, outliers
   
2. **Visualization**:
   - Library: Matplotlib, Seaborn, Plotly
   - Charts: Time series, bar charts, heatmaps, box plots
   
3. **Statistical Analysis**:
   - Hypothesis testing (uji komparatif 2023 vs 2024)
   - Correlation analysis antar produk
   - Regional comparison
   
4. **Machine Learning**:
   - Clustering: Segmentasi kabupaten berdasarkan pola konsumsi
   - Regression: Prediksi pengeluaran
   - Time series: Forecasting trend

---

**💡 Tips**: Setiap cell markdown dilengkapi dengan emoji untuk memudahkan navigasi visual!

In [59]:
print("="*70)
print(" "*20 + "📚 DOKUMENTASI NOTEBOOK 📚")
print("="*70)
print()
print("✅ Semua tahapan preprocessing telah dilengkapi dengan penjelasan detail!")
print()
print("📝 PENJELASAN YANG DITAMBAHKAN:")
print("-" * 70)
print("1. ✓ Import Libraries - Fungsi dan kegunaan setiap library")
print("2. ✓ Data Loading - Teknik custom loading untuk multi-level headers")
print("3. ✓ Data Cleaning - Strategi handling missing values & duplicates")
print("4. ✓ Data Transformation - Wide vs Long format explained")
print("5. ✓ Data Integration - Proses penggabungan datasets")
print("6. ✓ Feature Engineering - Pembuatan fitur turunan")
print("7. ✓ Quality Check - Validasi komprehensif")
print("8. ✓ Export - Use case setiap file output")
print()
print("="*70)
print()
print("📖 CARA MENGGUNAKAN NOTEBOOK INI:")
print("-" * 70)
print("• Markdown cells (biru) = Penjelasan konsep & metodologi")
print("• Code cells (abu-abu) = Implementasi dengan inline comments")
print("• Output cells = Hasil validasi")
print()
print("💡 TIPS:")
print("-" * 70)
print("• Scroll ke atas untuk membaca penjelasan dari awal")
print("• Setiap section memiliki emoji untuk navigasi mudah")
print("• Gunakan Ctrl+F untuk search keyword tertentu")
print()
print("="*70)
print(" "*15 + "🎉 Happy Learning & Analyzing! 🎉")
print("="*70)

                    📚 DOKUMENTASI NOTEBOOK 📚

✅ Semua tahapan preprocessing telah dilengkapi dengan penjelasan detail!

📝 PENJELASAN YANG DITAMBAHKAN:
----------------------------------------------------------------------
1. ✓ Import Libraries - Fungsi dan kegunaan setiap library
2. ✓ Data Loading - Teknik custom loading untuk multi-level headers
3. ✓ Data Cleaning - Strategi handling missing values & duplicates
4. ✓ Data Transformation - Wide vs Long format explained
5. ✓ Data Integration - Proses penggabungan datasets
6. ✓ Feature Engineering - Pembuatan fitur turunan
7. ✓ Quality Check - Validasi komprehensif
8. ✓ Export - Use case setiap file output


📖 CARA MENGGUNAKAN NOTEBOOK INI:
----------------------------------------------------------------------
• Markdown cells (biru) = Penjelasan konsep & metodologi
• Code cells (abu-abu) = Implementasi dengan inline comments
• Output cells = Hasil validasi

💡 TIPS:
----------------------------------------------------------------------
• 